File: ui/testing.py
import logging
import json
import time
import asyncio
from fastapi import Response
from nicegui import app

logger = logging.getLogger("ZenAI.Test")


def register_test_endpoints():
    """Register hidden testing endpoints for Monkey Test and chaos testing."""

    @app.post("/test/click/{element_id}")
    async def test_click_element(element_id: str):
        """Simulate a click on a UI element for chaos testing."""
        logger.info(f"[Test] Simulating click on: {element_id}")
        try:
            # Broadcast click to all connected clients
            clients = app.clients
            if callable(clients):
                try:
                    clients = clients()
                except TypeError:
                    pass

            # Support both dict and list-like clients
            client_list = []
            if isinstance(clients, dict):
                client_list = list(clients.values())
            elif clients:
                try:
                    client_list = list(clients)
                except (TypeError, ValueError):
                    client_list = []

            for client in client_list:
                try:
                    client.run_javascript(f"document.getElementById('{element_id}')?.click()")
                except Exception as e:
                    logger.debug(f"[Test] Failed for client: {e}")
                    continue
            return Response(status_code=200)
        except Exception as e:
            logger.error(f"[Test] Click simulation failed: {e}")
            return Response(content=str(e), status_code=500)

    @app.get("/test/state")
    async def test_get_ui_state():
        """Retrieve a summary of the current UI state (text and visible components)."""
        return {"active": True, "timestamp": time.time()}

    @app.post("/test/send")
    async def test_send_message(data: dict):
        """Simulate typing and sending a message for automated testing."""
        text = data.get("text", "")
        logger.info(f"[Test] Simulating send: {text}")
        try:
            # Broadcast to all clients (Super Robust)
            clients_obj = getattr(app, "clients", None)

            client_list = []
            if clients_obj:
                if callable(clients_obj):
                    try:
                        clients_obj = clients_obj()
                    except TypeError:
                        pass

                if isinstance(clients_obj, dict):
                    client_list = list(clients_obj.values())
                elif isinstance(clients_obj, (list, set, tuple)):
                    client_list = list(clients_obj)

            # Fallback to NiceGUI globals if app.clients failed
            if not client_list:
                try:
                    from nicegui import globals as ng_globals

                    client_list = list(ng_globals.clients.values())
                except (ImportError, AttributeError, TypeError):
                    pass

            logger.info(f"[Test] Broadasting to {len(client_list)} clients")

            for client in client_list:
                try:
                    js = f"""
                    (function() {{
                        const input = document.getElementById('ui-input-chat');
                        const btn = document.getElementById('ui-btn-send');
                        if (input) {{
                            input.value = {json.dumps(text)};
                            input.dispatchEvent(new Event('input', {{ bubbles: true }}));
                            console.log('[Test] Input set to:', input.value);
                            if (btn) {{
                                setTimeout(() => {{
                                    console.log('[Test] Clicking send button');
                                    btn.click();
                                }}, 100);
                            }}
                        }}
                    }})();
                    """
                    client.run_javascript(js)
                except Exception as e:
                    logger.debug(f"[Test] Send failed for client: {e}")
            return {"status": "ok"}
        except Exception as e:
            logger.error(f"[Test] Send simulation failed: {e}")
            return Response(content=str(e), status_code=500)

File: rust_output/src/testing.rs
use anyhow::{Result, Context};
use std::collections::HashMap;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Register hidden testing endpoints for Monkey Test and chaos testing.
pub fn register_test_endpoints() -> Result<()> {
    // Register hidden testing endpoints for Monkey Test and chaos testing.
    let test_click_element = |element_id| {
        // Simulate a click on a UI element for chaos testing.
        logger.info(format!("[Test] Simulating click on: {}", element_id));
        // try:
        {
            let mut clients = app::clients;
            if callable(clients) {
                // try:
                {
                    let mut clients = clients();
                }
                // except TypeError as _e:
            }
            let mut client_list = vec![];
            if /* /* isinstance(clients, dict) */ */ true {
                let mut client_list = clients.values().into_iter().collect::<Vec<_>>();
            } else if clients {
                // try:
                {
                    let mut client_list = clients.into_iter().collect::<Vec<_>>();
                }
                // except (TypeError, ValueError) as _e:
            }
            for client in client_list.iter() {
                // try:
                {
                    client.run_javascript(format!("document.getElementById('{}')?.click()", element_id));
                }
                // except Exception as e:
            }
            Response(/* status_code= */ 200)
        }
        // except Exception as e:
    };
    let test_get_ui_state = || {
        // Retrieve a summary of the current UI state (text and visible components).
        HashMap::from([("active".to_string(), true), ("timestamp".to_string(), std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64())])
    };
    let test_send_message = |data| {
        // Simulate typing and sending a message for automated testing.
        let mut text = data.get(&"text".to_string()).cloned().unwrap_or("".to_string());
        logger.info(format!("[Test] Simulating send: {}", text));
        // try:
        {
            let mut clients_obj = /* getattr */ None;
            let mut client_list = vec![];
            if clients_obj {
                if callable(clients_obj) {
                    // try:
                    {
                        let mut clients_obj = clients_obj();
                    }
                    // except TypeError as _e:
                }
                if /* /* isinstance(clients_obj, dict) */ */ true {
                    let mut client_list = clients_obj.values().into_iter().collect::<Vec<_>>();
                } else if /* /* isinstance(clients_obj, (list, set, tuple) */) */ true {
                    let mut client_list = clients_obj.into_iter().collect::<Vec<_>>();
                }
            }
            if !client_list {
                // try:
                {
                    // TODO: from nicegui import globals as ng_globals
                    let mut client_list = ng_globals.clients.values().into_iter().collect::<Vec<_>>();
                }
                // except (ImportError, AttributeError, TypeError) as _e:
            }
            logger.info(format!("[Test] Broadasting to {} clients", client_list.len()));
            for client in client_list.iter() {
                // try:
                {
                    let mut js = format!("\n                    (function() {{\n                        const input = document.getElementById('ui-input-chat');\n                        const btn = document.getElementById('ui-btn-send');\n                        if (input) {{\n                            input.value = {};\n                            input.dispatchEvent(new Event('input', {{ bubbles: true }}));\n                            console.log('[Test] Input set to:', input.value);\n                            if (btn) {{\n                                setTimeout(() => {{\n                                    console.log('[Test] Clicking send button');\n                                    btn.click();\n                                }}, 100);\n                            }}\n                        }}\n                    }})();\n                    ", serde_json::to_string(&text).unwrap());
                    client.run_javascript(js);
                }
                // except Exception as e:
            }
            HashMap::from([("status".to_string(), "ok".to_string())])
        }
        // except Exception as e:
    Ok(})
}
